1단계 미션 제출합니다.

Gongmeda opened the pull request.

findstar reviewed (approved):
안녕하세요~ 크리스. 리뷰어 코멘트 드리게된 제임스 입니다  😀
기능 구현이 잘되어 있고 단위 테스트도 확인했습니다. 
다만 몇가지 코멘트 남겼는데 확인 부탁드립니다. 
코멘트는 필수사항은 아니어서 Approve 했습니다.  💯 💯

아, 단위 테스트에서 ContentType fromUrlPath 메서드가 추가되면 좋겠습니다. 
다음 PR에서 챙겨주세요~
PR 하시느라 수고많으셨습니다 💯

kriery commented on src/test/java/webserver/model/HttpRequestTest.java:
```diff
@@ -88,7 +57,7 @@ void createUserWithPost() throws IOException {
         };
         InputStream in = new ByteArrayInputStream(String.join("\r\n", requestLines).getBytes());
-        Assertions.assertThat(Request.from(in).getParameters())
+        Assertions.assertThat(HttpRequest.from(in).getParameters())
```
수정: d553d2b

kriery commented on src/test/java/webserver/model/HttpRequestTest.java:
```diff
@@ -60,15 +29,15 @@ void parameters(int expected, String requestLine) throws IOException {
     }
 
     @Test
     void createUser() throws IOException {
```
수정: b4a5693

Gongmeda committed:
docs: README에 요구사항 정리

Gongmeda committed:
test: RestTemplate을 활용한 요청 테스트 작성

Gongmeda committed:
feat: index.html 기능 구현

Gongmeda committed:
refactor: index api 리팩터링

Gongmeda committed:
fix: 테스트 코드 오류 수정

Gongmeda committed:
refactor: HttpResponse로 응답 책임 이동

Gongmeda committed:
chore: HttpRequest headers 타입 수정

Gongmeda committed:
feat: queryParam으로 유저 생성 기능 구현

Gongmeda committed:
feat: POST user/create 구현

Gongmeda committed:
feat: redirect 응답 구현

Gongmeda committed:
refactor: 클래스 분리 및 코드 정리

Gongmeda committed:
docs: README 요구사항 체크

Gongmeda committed:
refactor: 테스트 코드 RestTemplate 중복 제거

sah3122 commented on src/main/java/controller/RequestController.java:
```diff
@@ -0,0 +1,123 @@
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+public class RequestController {
```
RequestController 가 너무 많은 역할을 하고 있는것 같습니다. 
Request에 해당하는 handler 를 매핑해주는 클래스와 비즈니스 로직을 수행하는 클래스를 분리해보는건 어떨까요 ?

Gongmeda replied:
반영했습니다!
https://github.com/next-step/jwp-was-kakao/pull/22/commits/c32bac129e92e2165351f09b2148b381d096d4a0

sah3122 commented on src/main/java/http/HttpHeaders.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.HashMap;
+
+public class HttpHeaders extends HashMap<String, String> {
```
HashMap Collection을 상속하는것과 조합을 사용하는것에 대해 차이점을 알고 계실까요 ? 
아래 문서를 통해 어떤 차이가 있는지 학습해보세요 😄 
https://tecoble.techcourse.co.kr/post/2020-05-18-inheritance-vs-composition/

Gongmeda replied:
반영했습니다!

sah3122 commented on src/main/java/http/StartLine.java:
```diff
@@ -0,0 +1,38 @@
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
+public class StartLine {
```
StartLine으로 정의하는것도 좋지만 일반적으로 `RequestLine`이라는 용어를 사용하기도 합니다 😄

Gongmeda replied:
용어 변경 반영했습니다!

sah3122 commented on src/main/java/controller/RequestController.java:
```diff
@@ -0,0 +1,123 @@
+        User user = new User(userId, password, name, email);
+
+        DataBase.addUser(user);
+        HttpResponse response = new HttpResponse();
```
Response 를 생성하기위한 정적팩토리 메서드를 정의해보는건 어떨까요 ? 
반복되는 코드를 줄일수 있을것 같아요 😄

Gongmeda replied:
반영했습니다!

sah3122 commented on src/main/java/controller/RequestController.java:
```diff
@@ -0,0 +1,123 @@
+        handleFileResponse("static", httpRequest, dos);
+    }
+
+    private static void handleRoot(DataOutputStream dos) throws IOException {
```
해당 함수는 request path가 정의되지 않은 요청을 처리하는 용도 일까요 ? 
맞다면 NotFound 와 같은 응답을 내려줘도 될것 같아요 😄

Gongmeda replied:
별도로 처리되지 않는 경로는 404 Not Found로 응답되도록 수정했습니다!

sah3122 commented on src/main/java/controller/RequestController.java:
```diff
@@ -0,0 +1,123 @@
+                return;
+            }
+            if (httpRequest.getStartLine().getMethod().equals(HttpMethod.POST)) {
+                String body = (String) httpRequest.getBody();
```
WebServer는 모든 요청에 대응할수 있는 기능을 제공해야 합니다. 
Content-Type에 따라 Request Body를 다루어 보는건 어떨까요 ?

sah3122 reviewed (changes requested):
1단계 미션 깔끔하게 구현 잘해주셨습니다 👍 
몇가지 생각거리를 남겨두었는데 확인 부탁드려요 🙏

Gongmeda committed:
refactor: HttpHeaders 조합으로 수정 및 http 객체 생성자 String으로 받는 로직 parser 쪽으로 분리

Gongmeda committed:
refactor: StartLine -> RequestLine

Gongmeda committed:
refactor: MimeParser 분리

Gongmeda committed:
refactor: 전반적인 리팩토링

- request, response 객체 파싱 책임 해당 객체로 이동 + Parser 제거
- request, response 객체 필드용 내부 객체 세분화 (HttpStatus, HttpMethod...)
- RequestMappingHandler 및 Controller 구현
- UserController 생성 (구현 TODO)

Gongmeda committed:
fix: ContentType에서 발생하는 NullPointerException 해결

Gongmeda committed:
refactor: URI 파싱 로직 수정

Gongmeda committed:
refactor: HttpHeaders add 메소드 추가 및 불변성 제거

Gongmeda committed:
feat: HttpStatus 404 추가

Gongmeda committed:
refactor: Controller 분리

Gongmeda committed:
fix: early return 부재로 인한 버그 해결

Gongmeda commented:
전반적으로 Spring MVC 구조와 유사하게 대규모 리팩토링 진행했습니다!
리팩토링 커밋 하나가 너무 크고 단일 PR에 적용된 변경이 너무 많이진 점은 양해 부탁드리겠습니다...ㅠ

Gongmeda committed:
refactor: 기본 응답에 대한 중복 코드 제거

Gongmeda committed:
test: URITest 구현

Gongmeda committed:
test: ContentTypeTest 구현

Gongmeda committed:
test: HttpHeadersTest 구현

Gongmeda committed:
test: StatusLineTest 구현

Gongmeda committed:
test: QueryParamsTest 구현

Gongmeda committed:
refactor: HttpRequestTest 정리

sah3122 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,33 @@
+import webserver.http.request.HttpRequest;
+import webserver.http.response.HttpResponse;
+
+public abstract class Controller {
```
Http 메서드에 따라 적절한 행위를 할수 있도록 추상 클래스를 정의해주셨네요 👍

sah3122 commented on src/main/java/controller/DefaultController.java:
```diff
@@ -0,0 +1,18 @@
+import webserver.http.request.HttpRequest;
+import webserver.http.response.HttpResponse;
+
+public class DefaultController extends Controller {
```
Default 라는 네이밍 보단 NotFoundController라는 네이밍도 어울려 보입니다 😄

sah3122 commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+        DataBase.addUser(user);
+        logger.info("add user: {}", user);
+
+        httpResponse.getHeaders().add("Location", "/index.html");
+        httpResponse.getStatusLine().setStatus(HttpStatus.FOUND);
```
response를 정의하는 코드는 해당 메서드에서 분리하는것이 좋아 보입니다 😄

sah3122 commented on src/main/java/webserver/http/ContentType.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Objects;
+import org.springframework.util.StringUtils;
+
+public enum ContentType {
```
Content Type 정의 👍

sah3122 commented on src/main/java/webserver/http/URI.java:
```diff
@@ -0,0 +1,29 @@
+package webserver.http;
```
http 스펙에 맞게 다양한 클래스를 정의해주셨네요 😄

sah3122 reviewed (approved):
기존 코드와 비교하여 많은 부분이 리팩토링 되었네요 👍 
다양한 클래스를 정의하고, 구조적으로 변경점이 많았는데, 리팩토링을 진행하는 과정에서 좋은 경험이 되었길 바랍니다 😄 
다음 미션도 화이팅 하세요 !
